src/upgrader.js

```diff
@@ -0,0 +1,359 @@
+        log('inbound: incoming stream opened on %s', protocol)
+        connection.addStream(stream, protocol)
+        this._onStream({ stream, protocol })
+      },
+      // Run anytime a stream ends
+      onStreamEnd: muxedStream => {
+        connection.removeStream(muxedStream.id)
+      }
+    })
+
+    // Called whenever a user requests a new stream
+    const newStream = async protocols => {
+      log('inbound: starting new stream on %s', protocols)
+      const muxedStream = muxer.newStream()
+      const mss = new Multistream.Dialer(muxedStream)
+      try {
+        const { stream, protocol } = await mss.select(protocols)
+        return { stream: { ...muxedStream, ...stream }, protocol }
+      } catch (err) {
+        log.error('could not create new stream for protocols %s', protocols, err)
+        throw errCode(err, codes.ERR_UNSUPPORTED_PROTOCOL)
+      }
+    }
+
+    // Pipe all data through the muxer
+    pipe(muxedConnection, muxer, muxedConnection)
+
+    maConn.timeline.upgraded = Date.now()
+    const timelineProxy = new Proxy(maConn.timeline, {
+      set: (...args) => {
+        if (args[1] === 'close' && args[2]) {
+          this.onConnectionEnd(connection)
+        }
+
+        return Reflect.set(...args)
+      }
+    })
+
+    // Create the connection
+    const connection = new Connection({
+      localAddr: maConn.localAddr,
+      remoteAddr: maConn.remoteAddr,
+      localPeer: this.localPeer,
+      remotePeer,
+      stat: {
+        direction: 'inbound',
+        timeline: timelineProxy,
+        multiplexer: Muxer.multicodec,
+        encryption: cryptoProtocol
+      },
+      newStream,
+      getStreams: () => muxer.streams,
+      close: err => maConn.close(err)
+    })
+
+    this.onConnection(connection)
+
+    return connection
+  }
+
+  /**
+   * Upgrades an outbound connection
+   * @async
+   * @param {MultiaddrConnection} maConn
+   * @returns {Promise<Connection>}
+   */
+  async upgradeOutbound (maConn) {
+    let remotePeerId
+    try {
+      remotePeerId = PeerId.createFromB58String(maConn.remoteAddr.getPeerId())
+    } catch (err) {
+      log.error('multiaddr did not contain a valid peer id', err)
+    }
+
+    let encryptedConn
+    let remotePeer
+    let muxedConnection
+    let cryptoProtocol
+    let Muxer
+
+    try {
+      // Encrypt the connection
+      ({
+        conn: encryptedConn,
+        remotePeer,
+        protocol: cryptoProtocol
+      } = await this._encryptOutbound(this.localPeer, maConn, remotePeerId, this.cryptos))
+
+      // Multiplex the connection
+      ;({ stream: muxedConnection, Muxer } = await this._multiplexOutbound(encryptedConn, this.muxers))
+    } catch (err) {
+      log.error('Failed to upgrade outbound connection', err)
+      await maConn.close(err)
+      throw err
+    }
+
+    log('Successfully upgraded outbound connection')
+
+    // Create the muxer
+    const muxer = new Muxer({
```

Comment:
what do you think of creating a function to create the muxer for outbound and inbound? It seems that only the log is different, so it would be easier to maintain less code

Reply:
multistream usage is different also, listener vs dialer and `select` vs `handle`

Reply:
From what I am seeing, I don't see those differences:

```js
// UpgradeInbound

  const muxer = new Muxer({
    // Run anytime a remote stream is opened
    onStream: async muxedStream => {
      const mss = new Multistream.Listener(muxedStream)
      const { stream, protocol } = await mss.handle(Array.from(this.protocols.keys()))
      log('inbound: incoming stream opened on %s', protocol)
      connection.addStream(stream, protocol)
      this._onStream({ stream, protocol })
    },
    // Run anytime a stream ends
    onStreamEnd: muxedStream => {
      connection.removeStream(muxedStream.id)
    }
  })

```

```js
// UpgradeOutbound
  const muxer = new Muxer({
    // Run anytime a remote stream is created
    onStream: async muxedStream => {
      const mss = new Multistream.Listener(muxedStream)
      const { stream, protocol } = await mss.handle(Array.from(this.protocols.keys()))
      log('outbound: incoming stream opened on %s', protocol)
      connection.addStream(stream, protocol)
      this._onStream({ stream, protocol })
    },
    // Run anytime a stream closes
    onStreamEnd: muxedStream => {
      connection.removeStream(muxedStream.id)
    }
  })

```

Same for `const newStream`

Reply:
Ah yes, you're right, I was looking at the wrong thing.

Reply:
The main issue with moving the muxer creation into a new function is that it's heavily reliant on the scope it's currently in. For example, it needs access to the connection, which can't be created until the muxer is made.

However, I do think we could move the whole of connection creation into its own function, which would keep the scope we need and do the deduplication.